### PR TITLE
Quaternion Networks - Bug Fixes & Improvements

### PR DESCRIPTION
**Pull request by Drew-Wagner**
## What does this PR do?
This PR fixes several bugs which prevented the use of the quaternion network modules, and completes the collection by implementing avg and max pooling. No tests existed for quaternion networks. This PR introduces a minimum (and incomplete) set of tests.

Several bugs were present in the existing quaternion network modules and are fixed by this PR:
- Bias was uninitialized for Conv2d and Conv1d when bias=False was set, leading to nan values in forward and / or backward pass.
- QBatchNorm training implementation did not match evaluation implementation: the mean was not being subtracted from the input.
- In QBatchNorm, a view was being used incorrectly to broadcast tensors together.
- Performance issue with QBatchNorm: torch.rsqrt is several times faster than 1 / torch.sqrt (I measured between a 3-8x speedup)
- Groups for QConv modules were broken: the input_channels for the weights must be divided by the # of groups.

Several adjustments were made to improve compatibility of the QConv interface with regular convolution modules:
- The `swap` option was added for QConv2d
- `max_norm` option was added for QConv and QLinear modules
- The bias parameter / buffer was renamed from `.b` to `.bias`

A QPooling2d module is added which implements:
- component-wise average pooling
- max pooling by magnitude of the quaternions

Breaking Changes:
- The bias parameter of QConv and QLinear modules was renamed from `.b` to `.bias`, however given the number of bugs present, it seems unlikely that anyone was depending on this.
- The # of input & output channels for QConv modules must be divisible by the # of groups.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/nnet/quaternion_networks/q_CNN.py [hunk 1]
```diff
@@ -189,6 +195,21 @@ def forward(self, x):
         """
         # (batch, channel, time)
         x = x.transpose(1, -1)
+
+        if self.max_norm is not None:
```
Comment (TParcollet) [inline speechbrain/nnet/quaternion_networks/q_CNN.py:199]: is renorming individual quaternion components actually is strictly equivalent to renorming the quaternion?

File: speechbrain/nnet/quaternion_networks/q_CNN.py [hunk 2]
```diff
@@ -363,6 +390,12 @@ class QConv2d(torch.nn.Module):
         with deep configurations. The vector_scale parameters are learnable
         parameters that acts like gates by multiplying the output vector with
         a small trainable parameter (default False).
+    max_norm: float
+        kernel max-norm.
+    swap: bool
+        If True, the convolution is done with the format (B, C, W, H).
+        If False, the convolution is dine with (B, H, W, C).
```
Comment (TParcollet) [inline speechbrain/nnet/quaternion_networks/q_CNN.py:397]: dine -> done

Comment (TParcollet) [reply speechbrain/nnet/quaternion_networks/q_CNN.py:397]: How does this relate to skip_transpose?

File: speechbrain/nnet/quaternion_networks/q_normalization.py [hunk 3]
```diff
@@ -120,38 +121,36 @@ def forward(self, input):
...
         else:
-            q_var = torch.cat(
-                [
-                    self.running_var,
-                    self.running_var,
-                    self.running_var,
-                    self.running_var,
-                ],
+            denominator = torch.rsqrt(self.running_var + self.eps)
+            denominator = torch.cat(
```
Comment (TParcollet) [inline speechbrain/nnet/quaternion_networks/q_normalization.py:149]: Wouldn't torch.repeat be faster?

**Review by TParcollet — COMMENTED** [review]
Thanks! LGTM, see my questions.

File: speechbrain/nnet/quaternion_networks/q_normalization.py (hunk 3 above)
Comment (Drew-Wagner) [reply speechbrain/nnet/quaternion_networks/q_normalization.py:149]: Yes it probably would be. I was just basing it off how the existing code was written, but I can go back and fix that this weekend.

File: speechbrain/nnet/quaternion_networks/q_CNN.py (hunk 2 above)
Comment (Drew-Wagner) [reply speechbrain/nnet/quaternion_networks/q_CNN.py:397]: Good point, I could add the skip_transpose option for better consistency with the regular Conv2d layers.

File: speechbrain/nnet/quaternion_networks/q_CNN.py (hunk 1 above)
Comment (Drew-Wagner) [reply speechbrain/nnet/quaternion_networks/q_CNN.py:199]: No I don't believe it is strictly equivalent. I couldn't find any references for how to approach it so I went with the simplest idea which was the component-wise renorm.

**Comment by mravanelli** [comment]
@Drew-Wagner, could you please fix the conflicts, merge the latest development, and do the last modifications?

File: speechbrain/nnet/quaternion_networks/q_ops.py [hunk 4]
```diff
@@ -853,3 +853,34 @@ def check_quaternion_input(input_shape):
...
+    )
+    renormed_weight_magnitude = torch.renorm(
+        weight_magnitude, p=2, dim=0, maxnorm=max_norm
+    )
+    factor = renormed_weight_magnitude / weight_magnitude
+
+    r_weight.data *= factor
+    i_weight.data *= factor
+    j_weight.data *= factor
+    k_weight.data *= factor
```
Comment (Drew-Wagner) [inline speechbrain/nnet/quaternion_networks/q_ops.py:886]: @TParcollet Please review this implementation which renorms the weights according to the magnitude of the quaternions, rather than by individual components

**Review by mravanelli — APPROVED** [review]

**Comment by mravanelli** [comment]
Thank you @Drew-Wagner for these fixes!
